File: HypernymExtractor/src/main/java/cz/vse/lhd/hypernymextractor/builder/HypernymExtractor.java
package cz.vse.lhd.hypernymextractor.builder;

import gate.Annotation;
import gate.AnnotationSet;
import gate.Corpus;
import gate.Document;
import gate.Factory;
import gate.FeatureMap;
import gate.Gate;
import gate.Node;
import gate.ProcessingResource;
import gate.creole.ExecutionException;
import gate.creole.ResourceInstantiationException;
import gate.creole.SerialAnalyserController;
import gate.util.GateException;
import gate.util.InvalidOffsetException;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.net.MalformedURLException;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.apache.commons.lang.StringEscapeUtils;

/**
 * @author Milan Dojchinovski
 * <milan (at) dojchinovski (dot) mk>
 * Twitter:
 * @m1ci www: http://dojchinovski.mk
 */
public class HypernymExtractor {

    private ProcessStatus loader;
    private SerialAnalyserController pipeline = null;
    private SerialAnalyserController resetPipeline = null;
    private final String hypernymLoggingPath;
    boolean prepared = false;
    private final boolean saveInTriplets;
    private String lang = "en";
    public String JAPEPATH;
    private final String taggerBinary_DE;
    private final String taggerBinary_NL;
    private final PrintWriter dbpediaOutputFile;
    private final PrintWriter rawOutputFile;
    private final DBpediaLinker dbpediaLinker;

    public HypernymExtractor(DBpediaLinker dbpediaLinker, String lang, String JAPEPATH, String hypernymLoggingPath, String taggerBinary_DE, String taggerBinary_NL) throws FileNotFoundException, GateException, MalformedURLException {
        this.lang = lang;
        this.JAPEPATH = JAPEPATH;
        this.saveInTriplets = saveInTriplets;
        this.hypernymLoggingPath = hypernymLoggingPath;
        this.taggerBinary_DE = taggerBinary_DE;
        this.taggerBinary_NL = taggerBinary_NL;
        dbpediaOutputFile = new PrintWriter(hypernymLoggingPath + ".dbpedia");
        rawOutputFile = new PrintWriter(hypernymLoggingPath + ".raw");
        this.dbpediaLinker = dbpediaLinker;
        Gate.getCreoleRegister().registerDirectories(
                new File(Gate.getPluginsHome(), "Tagger_Framework").toURL());
    }

    public void close() {
        dbpediaOutputFile.close();
        rawOutputFile.close();
    }

    public void setLoader(ProcessStatus loader) {
        this.loader = loader;
    }

    public ProcessStatus getLoader() {
        return loader;
    }

    public void extractHypernyms(Corpus corpus) {
        if (!prepared) {
            preparePipeline();
        }

        try {
            pipeline.setCorpus(corpus);
            pipeline.execute();
            for (Document doc : corpus) {
                loader.tryPrint();
                loader = loader.plusplus();
                AnnotationSet as_all = doc.getAnnotations();

                //                if (!ann_iter.hasNext()) {
//                    String articleName = (String) doc.getFeatures().get("article_title");
//                    Logger.getGlobal().log(Level.WARNING, "Article ''{0}'' no annotation ''h'' found.", articleName);
//                    //MyLogger.log(articleName + " ;NO Hypernym found");
//                }
                AnnotationSet as_hearst = as_all.get("h");
                for (Annotation isaAnnot : as_hearst) {
                    Node isaStart = isaAnnot.getStartNode();
                    Node isaEnd = isaAnnot.getEndNode();
                    String hypernym = doc.getContent().getContent(isaStart.getOffset(), isaEnd.getOffset()).toString();
                    
                    // override with lemma feature if present
                    if (isaAnnot.getFeatures().containsKey("lemma")) {
                        String hypCand = isaAnnot.getFeatures().get("lemma").toString();
                        if (hypCand != null && hypCand.length() > 0 && !hypCand.equals("<unknown>")) {
                            hypernym = hypCand;
                        }
//                        if (!hypernym.equals(hypCand)) {
//                            Logger.getGlobal().log(Level.INFO, "Replacing hypernym in text {0} with its DIFFERENT lemma: {1}", new Object[]{hypernym, hypCand});
//                        }
                    }

//                    Logger.getGlobal().log(Level.INFO, "HYPERNYM: {0}", hypernym);
                    if (hypernymLoggingPath != null && !"".equals(hypernymLoggingPath)) {
                        String dbPediaURL;
                        if (saveInTriplets) {
                            String urlName = (String) doc.getFeatures().get("dbpedia_url");
                            String urlHypernym = dbpediaLinker.getLink(hypernym);
                            if (urlHypernym == null) {
//                                Logger.getGlobal().log(Level.WARNING, "Hypernym found and not mapped: {0}", hypernym);
                            } else {
                                saveHypernym(urlName, urlHypernym, true);
                                doc.getFeatures().put("thdDBpediaType_search", urlName);
                            }
                        }
                        saveHypernym((String) doc.getFeatures().get("article_title"), hypernym, false);
                    } else {
                        //MyLogger.log((String) doc.getFeatures().get("article_title") + " ;Hypernym found");
                    }

                    //resetPipeline.setCorpus(corpus);
                    //resetPipeline.execute();
                    FeatureMap f = Factory.newFeatureMap();
                    //f.put("url", Linker.getInstance().getWikiLinkEN(hypernym,"local"));  
                    doc.getAnnotations().add(isaStart, isaEnd, "a", f);
                }
            }

        } catch (InvalidOffsetException ex) {
            Logger.getGlobal().log(Level.SEVERE, null, ex);
        } catch (ExecutionException ex) {
            Logger.getGlobal().log(Level.SEVERE, null, ex);
        }
    }

    private void saveHypernym(String source, String hypernym, boolean mapToDBpedia) {

        if (mapToDBpedia == false) {
            StringBuilder content = new StringBuilder();
            content.append(source).append(";").append(hypernym);
            rawOutputFile.println(content);
        } else {
            StringBuilder content = new StringBuilder();
            content.append("<");
            content.append(StringEscapeUtils.escapeJava(source));
            content.append(">");
            //content.append(" <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ");
            content.append(" <?> ");
            content.append("<");
            content.append(StringEscapeUtils.escapeJava(hypernym));
            content.append(">.");
            dbpediaOutputFile.println(content);
        }

//            File file = new File(concreteHypernymLoggingPath);
//
//            // if file doesnt exists, then create it
//            if (!file.exists()) {
//                file.createNewFile();
//            }
//
//            FileWriter fw = new FileWriter(file.getAbsoluteFile(), true);
//            new PrintWriter(file)
//            BufferedWriter bw = new BufferedWriter(fw);
//            bw.write(content.toString());
//            bw.close();
    }

    private void preparePipeline() {
        try {
            prepared = true;

            //pipeline = (SerialAnalyserController) Factory.createResource("gate.creole.SerialAnalyserController");
            pipeline = (SerialAnalyserController) Factory.createResource("gate.creole.RealtimeCorpusController");
            FeatureMap resetFeatureMap = Factory.newFeatureMap();
            ProcessingResource resetPR = (ProcessingResource) Factory.createResource("gate.creole.annotdelete.AnnotationDeletePR", resetFeatureMap);
            pipeline.add(resetPR);

            FeatureMap tokenizerFeatureMap = Factory.newFeatureMap();
            ProcessingResource tokenizerPR = (ProcessingResource) Factory.createResource("gate.creole.tokeniser.DefaultTokeniser", tokenizerFeatureMap);

            pipeline.add(tokenizerPR);

            if (lang.equals("en")) {
                FeatureMap sentenceSplitterFeatureMap = Factory.newFeatureMap();
                ProcessingResource sentenceSplitterPR = (ProcessingResource) Factory.createResource("gate.creole.splitter.RegexSentenceSplitter", sentenceSplitterFeatureMap);

                FeatureMap posTaggerFeatureMap = Factory.newFeatureMap();
                ProcessingResource posTaggerPR = (ProcessingResource) Factory.createResource("gate.creole.POSTagger", posTaggerFeatureMap);
                pipeline.add(sentenceSplitterPR);
                pipeline.add(posTaggerPR);
            } else if (lang.equals("nl") | lang.equals("de")) {

                FeatureMap taggerFeatureMap = Factory.newFeatureMap();
                taggerFeatureMap.put("debug", "false");
                taggerFeatureMap.put("encoding", "utf-8");
//                if (!lang.equals("de")) {
//                    taggerFeatureMap.put("encoding", "utf-8");
//                } else {
//                    // for de, utf-8 causes on some documents tagger to fail
//                    taggerFeatureMap.put("encoding", "ISO-8859-1");
//                }

                taggerFeatureMap.put("failOnUnmappableCharacter", "false");
                taggerFeatureMap.put("featureMapping", "lemma=3;category=2;string=1");
                taggerFeatureMap.put("inputAnnotationType", "Token");
                taggerFeatureMap.put("inputTemplate", "${string}");
                taggerFeatureMap.put("outputAnnotationType", "Token");
                taggerFeatureMap.put("regex", "(.+)	(.+)	(.+)");
                if (lang.equals("de")) {
                    taggerFeatureMap.put("taggerBinary", taggerBinary_DE);

                } else if (lang.equals("nl")) {
                    taggerFeatureMap.put("taggerBinary", taggerBinary_NL);
                }
                /*taggerFeatureMap.put("taggerFlags", "[]");
                 taggerFeatureMap.put("updateAnnotations", "true");
                            
                 taggerFeatureMap.put("preProcessURL", "");
                 taggerFeatureMap.put("postProcessURL", "");
                            
                 taggerFeatureMap.put("inputASName", "");
                 taggerFeatureMap.put("outputASName", "");
                 taggerFeatureMap.put("taggerDir", "");*/

                ProcessingResource genTag = (ProcessingResource) Factory.createResource("gate.taggerframework.GenericTagger", taggerFeatureMap);
                //genTag.setFeatures(taggerFeatureMap); 

                pipeline.add(genTag);
            }

            File japeOrigFile = new File(JAPEPATH);
            //File japeOrigFile = new File("/Users/Milan/Documents/Programming/repositories/linkedtv/WP2/THD/code/CorpusBuilderPR/en_hearst.jape");
            java.net.URI japeURI = japeOrigFile.toURI();
            FeatureMap transducerFeatureMap = Factory.newFeatureMap();
            try {
                transducerFeatureMap.put("grammarURL", japeURI.toURL());
                transducerFeatureMap.put("encoding", "UTF-8");
            } catch (MalformedURLException e) {
                System.out.println("Malformed URL of JAPE grammar");
                System.out.println(e.toString());
            }
            ProcessingResource japeCandidatesPR = (ProcessingResource) Factory.createResource("gate.creole.Transducer", transducerFeatureMap);

            pipeline.add(japeCandidatesPR);

            resetPipeline = (SerialAnalyserController) Factory.createResource("gate.creole.SerialAnalyserController");

            FeatureMap resetFeatureMap2 = Factory.newFeatureMap();
            ProcessingResource resetPR2 = (ProcessingResource) Factory.createResource("gate.creole.annotdelete.AnnotationDeletePR", resetFeatureMap2);

            resetPipeline.add(resetPR2);

        } catch (ResourceInstantiationException ex) {
            Logger.getGlobal().log(Level.SEVERE, null, ex);
        }
    }
}

File: HypernymExtractor/src/main/scala/cz/vse/lhd/hypernymextractor/builder/HypernymExtractor.scala
package cz.vse.lhd.hypernymextractor.builder

import java.io.File

import cz.vse.lhd.hypernymextractor.builder.HypernymExtractor._
import cz.vse.lhd.hypernymextractor.{Conf, ProcessStatus}
import gate.creole.SerialAnalyserController
import gate.{Corpus, Factory, ProcessingResource}
import org.slf4j.LoggerFactory

import scala.collection.JavaConverters._

/**
 * Created by propan on 17. 4. 2015.
 */
class HypernymExtractor private (dbpediaLinker: DBpediaLinker, private var processStatus: ProcessStatus) {

  private val lang = Conf.lang
  private lazy val logger = LoggerFactory.getLogger(getClass)

  private lazy val pipeline = {
    val pipeline = Factory.createResource("gate.creole.RealtimeCorpusController").asInstanceOf[SerialAnalyserController]
    val resetPR = Factory.createResource("gate.creole.annotdelete.AnnotationDeletePR", Factory.newFeatureMap()).asInstanceOf[ProcessingResource]
    val tokenizerPR = Factory.createResource("gate.creole.tokeniser.DefaultTokeniser", Factory.newFeatureMap()).asInstanceOf[ProcessingResource]
    pipeline.add(resetPR)
    pipeline.add(tokenizerPR)

    if (lang.equals("en")) {
      val sentenceSplitterPR = Factory.createResource("gate.creole.splitter.RegexSentenceSplitter", Factory.newFeatureMap()).asInstanceOf[ProcessingResource]
      val posTaggerPR = Factory.createResource("gate.creole.POSTagger", Factory.newFeatureMap()).asInstanceOf[ProcessingResource]
      pipeline.add(sentenceSplitterPR)
      pipeline.add(posTaggerPR)
    } else if (lang.equals("nl") || lang.equals("de")) {
      val taggerFeatureMap = Factory.newFeatureMap()
      taggerFeatureMap.put("debug", "false")
      taggerFeatureMap.put("encoding", "utf-8")
      //                if (!lang.equals("de")) {
      //                    taggerFeatureMap.put("encoding", "utf-8");
      //                } else {
      //                    // for de, utf-8 causes on some documents tagger to fail
      //                    taggerFeatureMap.put("encoding", "ISO-8859-1");
      //                }
      taggerFeatureMap.put("failOnUnmappableCharacter", "false")
      taggerFeatureMap.put("featureMapping", "lemma=3;category=2;string=1")
      taggerFeatureMap.put("inputAnnotationType", "Token")
      taggerFeatureMap.put("inputTemplate", "${string}")
      taggerFeatureMap.put("outputAnnotationType", "Token")
      taggerFeatureMap.put("regex", "(.+)	(.+)	(.+)")
      if (lang.equals("de")) {
        taggerFeatureMap.put("taggerBinary", "resources/TreeTagger/tree-tagger-german-gate")
      } else if (lang.equals("nl")) {
        taggerFeatureMap.put("taggerBinary", "resources/TreeTagger/tree-tagger-dutch-gate")
      }
      val genTag = Factory.createResource("gate.taggerframework.GenericTagger", taggerFeatureMap).asInstanceOf[ProcessingResource]
      pipeline.add(genTag)
    }
    val transducerFeatureMap = Factory.newFeatureMap()
    transducerFeatureMap.put("grammarURL", new File(Conf.gateJapeGrammar).toURI.toURL)
    transducerFeatureMap.put("encoding", "UTF-8")
    val japeCandidatesPR = Factory.createResource("gate.creole.Transducer", transducerFeatureMap).asInstanceOf[ProcessingResource]
    pipeline.add(japeCandidatesPR)
    pipeline
  }

  def extractHypernyms(corpus: Corpus)(implicit processHypernym: Hypernym => Unit) = {
    logger.debug(s"Corpus pipeline executing with size: ${corpus.size()}")
    pipeline.setCorpus(corpus)
    pipeline.execute()
    for (doc <- corpus.iterator().asScala) {
      processStatus.tryPrint()
      processStatus = processStatus.plusplus
      for (isaAnnot <- doc.getAnnotations.get("h").iterator().asScala) {
        val hypernym = {
          val hypernym = Option(isaAnnot.getFeatures)
            .filter(_.containsKey("lemma"))
            .map(_.get("lemma"))
            .filter(_ != null)
            .map(_.toString)
            .filter(x => x.length > 0 && x != "<unknown>")
            .getOrElse(doc.getContent.getContent(isaAnnot.getStartNode.getOffset, isaAnnot.getEndNode.getOffset).toString)
          Hypernym(
            doc.getFeatures.get("article_title").asInstanceOf[String],
            doc.getFeatures.get("dbpedia_url").asInstanceOf[String],
            hypernym,
            dbpediaLinker.getLink(hypernym)
          )
        }
        logger.debug(s"Hypernym extracted: $hypernym")
        processHypernym(hypernym)
      }
    }
  }

  /*  try {} catch {
      case e @ (_: InvalidOffsetException | _: ExecutionException) => logger.error(e.getMessage, e)
    }*/

}

object HypernymExtractor {

  case class Hypernym(resourceName: String, resourceUri: String, rawHypernym: String, resourceHypernym: Option[String])

  def apply(dbpediaLinker: DBpediaLinker, start: Int, end: Int)(extractAll: HypernymExtractor => Unit) = {
    extractAll(new HypernymExtractor(dbpediaLinker, new ProcessStatus(end - start)))
  }

}